### ISSUE TITLE
NumaPolicy issues regarding running multiple instances of Stockfish.

### ISSUE THREAD

**Issue by PikaCat-OuO** [official-stockfish/Stockfish#5599]
### Describe the issue

While tweaking Stockfish's `NumaPolicy` options, I noticed some unusual behavior. I'm unsure if this is by design or a bug, so I'm posting it here for further examination.

Let's define the following configurations:
- **MIMT** (Multiple Instances, Multiple Threads): Running multiple instances of Stockfish, each with multiple threads.
- **SIMT** (Single Instance, Multiple Threads): Running a single instance of Stockfish with multiple threads.
- **MIST** (Multiple Instances, Single Thread): Running multiple instances of Stockfish, each with a single thread.
- **SIST** (Single Instance, Single Thread): Running a single instance of Stockfish with a single thread.

The issues are as follows:
1. With NumaPolicy set to `auto`, the NPS (Nodes Per Second) of **MIMT** is noticeably lower than that of **SIMT**, even when resources are adequate for running multiple instances. This is evident when comparing Figure 2 and Figure 3.
2. In the **SIMT** scenario, the NPS with NumaPolicy set to `auto` is visibly lower than when set to `hardware`. This can be seen by comparing Figure 1 and Figure 3.
3. In the **MIST** scenario, the NPS with NumaPolicy set to `hardware` is significantly lower than when set to `auto`, even when resources are sufficient. This is shown by comparing Figure 4 and Figure 5.

Some of these issues may affect the effectiveness of engine testing, such as running the Fishtest client on a NUMA machine.

### Expected behavior

1. With NumaPolicy set to `auto`, the NPS of **MIMT** should be close to that of **SIMT** when resources are adequate for running multiple instances.
2. In the **SIMT** scenario, the NPS with NumaPolicy set to `auto` should be similar to when it's set to `hardware`.
3. In the **MIST** scenario, the NPS with NumaPolicy set to `hardware` should be comparable to when it's set to `auto`, provided resources are sufficient.

### Steps to reproduce

CPU: AMD EPYC 7T83 64-Cores x 2 sockets, SMT on.

![b6a97f74bc6dfc6789daa20974a7c344](https://github.com/user-attachments/assets/149c6ccd-91ed-4c7e-8d2b-9f6e7ed1755b)
Figure 1. Running a single Stockfish instance with 64 threads and NumaPolicy set to `hardware`.

![8edd46cffa72a32470071c150b2f5587](https://github.com/user-attachments/assets/93bd713d-a3cd-4420-842d-8c498cd17b0c)
Figure 2. Running two Stockfish instances each with 64 threads and NumaPolicy set to `auto`.

![2dfa6f4eb38cc319aa7499ca9f2e622d](https://github.com/user-attachments/assets/d915bcf7-3c47-47f5-99e3-62cc5a46f8ab)
Figure 3. Running a single Stockfish instance with 64 threads and NumaPolicy set to `auto`.

![df1f1e315f30c04889f5151f591e49cb](https://github.com/user-attachments/assets/c265872b-2dae-4aa6-8fde-fa80431cad97)
Figure 4. Running two Stockfish instances each with a single thread and NumaPolicy set to `hardware`.

![eeca4ba2a726dc1851602034c695e9c2](https://github.com/user-attachments/assets/6677b619-3274-4532-adde-b92dad880385)
Figure 5. Running two Stockfish instances each with a single thread and NumaPolicy set to `auto`.

### Operating system

Windows

### Stockfish version

f677aee2

**Comment by Sopel97** [comment]
Alright, this needs some breaking down...

1.

> With NumaPolicy set to auto, the NPS (Nodes Per Second) of MIMT is noticeably lower than that of SIMT

that should be true, yes, as expected, because the two instances would be competing for cache

> even when resources are adequate for running multiple instances. This is evident when comparing Figure 2 and Figure 3.

this is NOT what you're observing on figs 2 and 3. Figure 2 shows two stockfish instances, each running with 64 threads, but in different processor groups - effectively utilizing 128 threads. Figure 3 shows one stockfish instance utilizing 64 threads. The combined nps in Figure 2 is higher than in figure 3, which is a) not relevant to the claim from the previous quotation due to wrong assumptions, and b) opposite of what you described is happening.

--------------------

2.

> In the SIMT scenario, the NPS with NumaPolicy set to auto is visibly lower than when set to hardware. This can be seen by comparing Figure 1 and Figure 3.

A single processor group in your system consists of 64 threads but only 32 cores. Something is constraining your stockfish instance to run on a single processor group, which Stockfish is now NOT overriding unless explicitly informed by specifiying a custom, or `hardware` numa policy. See https://github.com/official-stockfish/Stockfish/issues/5307#issuecomment-2145217537.

In particular, when you use numa policy `hardware` the OS is free to schedule each thread to a separate physical core, so obviously there is will increase in speed.

-------------------

3. 

> In the MIST scenario, the NPS with NumaPolicy set to hardware is significantly lower than when set to auto, even when resources are sufficient. This is shown by comparing Figure 4 and Figure 5.

In each case Stockfish is running while bound to different processor groups. It just so happened that the first processor group was either slow or more occupied. It's common for these chips to have different CCDs able to boost differently.

----------------------

In other words, please investigate why processes you ran from the command line have their affinity restricted to a single processor group.

**Comment by Disservin** [comment]
> The combined nps in Figure 2 is higher than in figure 3, which is a) not relevant to the claim from the previous quotation due to wrong assumptions, and b) opposite of what you described is happening.

The system has 128 core, what they wanted to say is that when 2, 64 thread instances are started the nps of them individually are worse compared to when just 1 instance is running. Ideally the the threads should all be on their own core and not require hyperthreading, but since the affinity is somehow getting restricted this leads to not so great performance.

But this all leads to the question which you asked, why the affinity is somehow restricted.

**Comment by Sopel97** [comment]
They are both running on the same socket in which case the fixed power budget for the CPU will cause lower individual performance.

**Comment by PikaCat-OuO** [comment]
One interesting thing is that it only appears on Windows Server, and these issues are gone when running the engine on Windows 11. The system in the screenshots is Windows Server 2016.

**Comment by Disservin** [comment]
> please investigate why processes you run from the command line have their affinity restricted to a single processor group

Any idea what @PikaCat-OuO can do to figure that out ?
I think https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer lets you view the affinity for a process maybe you can use this to check what affinity the cli has.

Also it looks like you started Stockfish by double clicking the .exe file, can you try to invoke stockfish over cmd and check the results there, and if that fails check with terminal?

**Comment by PikaCat-OuO** [comment]
> Also it looks like you started Stockfish by double clicking the .exe file, can you try to invoke stockfish over cmd and check the results there, and if that fails check with terminal?

It may be involving some luck, I ran the program multiple times by double click and run from the terminal, here are my results:

![b3a2aef3771f3a20f860d327246b3d26](https://github.com/user-attachments/assets/556e285e-c2b0-4172-a1e0-542f9b9c14a8)
![9ed7c80ebf79fcf695b53cd930ead8eb](https://github.com/user-attachments/assets/e4cfec02-ccb4-4975-9657-455965dd31f4)
Figure 6. Eight different instances run from the command line.

![ae7f9005eee7d3b69db954e9fcfc3089](https://github.com/user-attachments/assets/fe4b9027-71f9-4a4c-8752-dddb9e13d38e)
![ac61a0d5366bd4448218d97081098bdb](https://github.com/user-attachments/assets/2ee362b3-c0d5-455c-a58f-feb9f29dc88c)
Figure 7. Eight different instances run from double click.

From here, I can see that it sometimes finds itself in different sockets, which may indicate that affinity does not get infected.

**Comment by Disservin** [comment]
That doesn’t really help, in all your cases it ends up in random processor group.
But the question is why it doesn’t see all processors with the auto setting. It should see all processors which are listed when the option hardware is set if I understood it correctly

**Comment by PikaCat-OuO** [comment]
Oh, I see what you mean, I originally thought you meant different instances always get scheduled to the same processor group.

**Comment by PikaCat-OuO** [comment]
![5f2004b2ef808f71a400289adfb6ef3f](https://github.com/user-attachments/assets/52c5e587-2977-4d02-a508-f3cfdce72ada)
With the tools, I can only see 64 processors. But skimming through the internet, I think I have the root cause.

![image](https://github.com/user-attachments/assets/ac493c8a-88b2-4ded-ba98-13c1f6778b44)
![image](https://github.com/user-attachments/assets/83c28be9-dbe4-43e6-b9db-c2f9083d2ffa)
https://learn.microsoft.com/zh-cn/windows/win32/procthread/processor-groups

This is likely a Windows Server issue. Don't know if this will be helpful:
![image](https://github.com/user-attachments/assets/9bf72b89-912a-496d-885f-c1aefedc8588)

**Comment by PikaCat-OuO** [comment]
Okay, I have more to say. If it's the case of Stockfish 16.1, even the processors available shown in the tools you provide is still 64:
![67a5dd958596684cd6171204917a8eff](https://github.com/user-attachments/assets/865bc150-04b8-4cba-a8a1-6d3f2b4df3d0)

When setting it to use 128 threads, it can still utilize more than 64 processors.
![1537dcdf27ea4f2389bcd39733f5c00d](https://github.com/user-attachments/assets/b4f34e4a-4895-4591-a5ad-f5cf6adc3d10)

Whereas with this new Stockfish, we lost this ability, and can only use 64 processors even if we set it to use 128 threads:
![27c5853d7e89ce38e41f3023646da14d](https://github.com/user-attachments/assets/4f05c53c-585e-4fbb-9ddd-cf881058921f)

So apparently something is still wrong in the NumaPolicy I guess.

**Comment by Disservin** [comment]
Well yes the numa option auto respects the affinity which is set from the parent process, whereas the hardware option ignores it and forces the use of all cores. When you start stockfish over a double click it should have no parent process which limits its available cores and as such it should use all cores. Now in your case this seems to be what is happening. This is why sopel said something restrict the affinity. 

Are you sure sopel that our current code behaves correctly, I mean could be that the retrieval just doesn’t work? Would explain why this here fails or why it fails in chessbase or in the tauri app ? Always seemed odd to be that they would somehow tinker with the affinity

**Comment by Sopel97** [comment]
> The system in the screenshots is Windows Server 2016.

Ahhh. Yes, this is a shortcoming of older Windows - initial process affinity is bound to a single processor group. It is however indistinguishable from affinity set by the user, so we must respect that. Previous Stockfish was incorrectly overriding this behaviour implicitly. Now it requires an explicit user setting.

**Comment by Disservin** [comment]
this could probably use some more documentation somewhere
